src/type-layout.md

```diff
@@ -0,0 +1,241 @@
+alignment of the field. If the current offset is not a multiple of the field's
+alignment, then add padding bytes increasing the current offset until the
+current offset is a multiple of the field's alignment. The offset for the field
+is what the current offset is now. Then increase the current offset by the size
+of the field.
+
+Finally, the size of the struct is the current offset rounded up to the nearest
+multiple of the struct's alignment. 
+
+> Note: You can have zero-sized structs from this algorithm. This differs from
+> C where structs without data still have a size of one byte.
+
+#### \#[repr(C)] Unions
+
+A union declared with `#[repr(C)]` will have the same size and alignment as an
+equivalent C union declaration in the C language for the target platform.
+Usually, a union would have the maximum size of the maximum size of all of its
+fields, and the maximum alignment of the maximum alignment of all of its fields.
+These maximums may come from different fields.
+
+```
+#[repr(C)]
+union Union {
+    f1: u16,
+    f2: [u8; 4],
+}
+
+assert_eq!(std::mem::size_of::<Union>(), 4);  // From f2
+assert_eq!(std::mem::align_of::<Union>(), 2); // From f1
+```
+
+#### \#[repr(C)] Enums
+
+For [C-like enumerations], the `C` representation has the size and alignment of
+the default `enum` size and alignment for the target platform's C ABI.
+
+> Note: The enum representation in C is implementation defined, so this is
+> really a "best guess". In particular, this may be incorrect when the C code
+> of interest is compiled with certain flags.
+
+> Warning: There are crucial differences between an `enum` in the C language and
+> Rust's C-like enumerations with this representation. An `enum` in  C is
+> mostly a `typedef` plus some named constants; in other words, an object of an
+> `enum` type can hold any integer value. For example, this is often used for
+> bitflags in `C`. In contrast, Rust’s C-like enumerations can only legally hold
+> the discrimnant values, everything else is undefined behaviour. Therefore, 
+> using a C-like enumeration in FFI to model a C `enum` is often wrong.
+
+It is an error for [zero-variant enumerations] to have the `C` representation.
+
+For all other enumerations, the layout is unspecified.
+
+### Primitive representations
+
+The *primitive representations* are the representations with the same names as
+the primitive integer types. That is: `u8`, `u16`, `u32`, `u64`, `usize`, `i8`,
+`i16`, `i32`, `i64`, and `isize`.
+
+Primitive representations can only be applied to enumerations.
+
+For [C-like enumerations], they set the size and alignment to be the same as the
+primitive type of the same name. For example, a C-like enumeration with a `u8`
+representation can only have discriminants between 0 and 255 inclusive.
+
+It is an error for [zero-variant enumerations] to have a primitive
+representation.
+
+For all other enumerations, the layout is unspecified.
```

Comment:
Also unspecified: `(C, primitive)`, `(primitive1, primitive2)`

Reply:
For now, I'm pretending you cannot do that. The first one will come in once your RFC is merged. The second one should probably be an error.

Reply:
It's currently only ever been a warning, so I think we should be clear here. Technically not back-compat to make an error.